File: packages/setpieces-r/setpieces/R/secondballs.R
# Second-ball contest detection after a team's own set pieces.
#
# For each qualifying delivery, look at the next events (within a short
# time window) for contested actions (aerial duel, tackle, interception,
# clearance, ball recovery). The *second* such contested action is
# treated as "the second ball" -- the first is usually just the target
# player's initial header/contact, the second is who actually comes
# away with the loose ball. This is very often well outside the box,
# since a strong clearance can travel 40+ metres before the second ball
# is won.

FINAL_THIRD_START <- 200.0 / 3.0

#' Second-ball contests following a team's own set pieces
#'
#' @param events `sorted_events(match)` for the same match `set_pieces`
#'   was built from -- contest detection scans forward from each
#'   delivery's `event_index` in this list.
#' @param set_pieces A list of set-piece event records (from
#'   [extract_set_pieces()]) for the same match.
#' @param team_id The delivering team's contestantId.
#' @param window_events Max number of subsequent events to scan.
#' @param window_minutes Max time (in minutes) after the delivery to
#'   keep scanning (default 12 seconds).
#' @param max_clean_passes_between Max number of clean (successful)
#'   passes allowed between the first and second contested action
#'   before treating the chain as broken.
#' @param final_third_start X-coordinate a delivery must reach to
#'   qualify (default: start of the final third).
#' @return A list of contest records, each with `delivery` (the
#'   originating set-piece event), `x`, `y`, `won` (logical), and
#'   `winner_player`.
#' @export
find_second_ball_contests <- function(events, set_pieces, team_id,
                                       window_events = 5, window_minutes = 12 / 60,
                                       max_clean_passes_between = 1,
                                       final_third_start = FINAL_THIRD_START) {
  results <- list()

  for (sp in set_pieces) {
    if (!identical(sp$team_id, team_id)) next
    if (is.na(sp$end_x) || sp$end_x < final_third_start) next

    i <- sp$event_index
    t0 <- minute_value(events[[i]])

    window <- list()
    j <- i + 1
    while (j <= length(events) && length(window) < window_events) {
      if (minute_value(events[[j]]) - t0 > window_minutes) break
      window[[length(window) + 1]] <- events[[j]]
      j <- j + 1
    }
    if (length(window) == 0) next

    first_idx <- NA_integer_
    for (k in seq_along(window)) {
      if (!is.null(window[[k]]$typeId) && window[[k]]$typeId %in% CONTESTED_TYPES) {
        first_idx <- k
        break
      }
    }
    if (is.na(first_idx)) next

    second_ball <- NULL
    clean_passes <- 0
    if (first_idx < length(window)) {
      for (k in (first_idx + 1):length(window)) {
        ev <- window[[k]]
        if (!is.null(ev$typeId) && ev$typeId %in% CONTESTED_TYPES) {
          second_ball <- ev
          break
        }
        if (!is.null(ev$typeId) && ev$typeId == TYPE_PASS && !is.null(ev$outcome) && ev$outcome == 1) {
          clean_passes <- clean_passes + 1
          if (clean_passes > max_clean_passes_between) break
        }
      }
    }
    if (is.null(second_ball) || is.null(second_ball$x) || is.null(second_ball$y)) next

    won <- identical(second_ball$contestantId, team_id)
    results[[length(results) + 1]] <- list(
      delivery = sp, x = as.numeric(second_ball$x), y = as.numeric(second_ball$y),
      won = won,
      winner_player = if (is.null(second_ball$playerName)) NA_character_ else second_ball$playerName
    )
  }

  results
}

File: packages/setpieces-py/src/setpieces/secondballs.py
"""Second-ball contest detection after a team's own set pieces.

For each qualifying delivery, look at the next events (within a short
time window) for contested actions (aerial duel, tackle, interception,
clearance, ball recovery). The *second* such contested action is treated
as "the second ball" -- the first is usually just the target player's
initial header/contact, the second is who actually comes away with the
loose ball. This is very often well outside the box, since a strong
clearance can travel 40+ metres before the second ball is won.
"""

from dataclasses import dataclass
from typing import List, Optional

from . import codes
from .events import SetPieceEvent
from .geometry import minute_value

FINAL_THIRD_START = 200.0 / 3.0


@dataclass(frozen=True)
class SecondBallContest:
    delivery: SetPieceEvent
    x: float
    y: float
    won: bool
    winner_player: Optional[str]


def find_second_ball_contests(events: List[dict], set_pieces: List[SetPieceEvent], team_id: str,
                               window_events: int = 5, window_minutes: float = 12 / 60,
                               max_clean_passes_between: int = 1,
                               final_third_start: float = FINAL_THIRD_START) -> List[SecondBallContest]:
    """Second-ball contests following ``team_id``'s own corners, free
    kicks, throw-ins and penalty-box direct free-kick shots.

    ``events`` must be ``setpieces.sorted_events(match)`` for the same
    match that ``set_pieces`` (``setpieces.extract_set_pieces(match)``)
    was built from -- contest detection scans forward from each
    delivery's ``event_index`` in that list.
    """
    results: List[SecondBallContest] = []

    for sp in set_pieces:
        if sp.team_id != team_id:
            continue
        if sp.end_x is None or sp.end_x < final_third_start:
            continue

        i = sp.event_index
        t0 = minute_value(events[i])
        window = []
        for ev in events[i + 1:i + 1 + window_events]:
            if minute_value(ev) - t0 > window_minutes:
                break
            window.append(ev)

        first_idx = next((j for j, ev in enumerate(window) if ev.get("typeId") in codes.CONTESTED_TYPES), None)
        if first_idx is None:
            continue

        second_ball, clean_passes = None, 0
        for ev in window[first_idx + 1:]:
            if ev.get("typeId") in codes.CONTESTED_TYPES:
                second_ball = ev
                break
            if ev.get("typeId") == codes.TYPE_PASS and ev.get("outcome") == 1:
                clean_passes += 1
                if clean_passes > max_clean_passes_between:
                    break

        if second_ball is None or second_ball.get("x") is None or second_ball.get("y") is None:
            continue

        won = second_ball.get("contestantId") == team_id
        results.append(SecondBallContest(
            delivery=sp, x=float(second_ball["x"]), y=float(second_ball["y"]),
            won=won, winner_player=second_ball.get("playerName"),
        ))

    return results
